### PR TITLE
Closes #2454: Add animations

### LINKED ISSUE #2454: Add animations
### Why/User Benefit/User Problem
Users are confused that the overlay is an overlay over the web content. Adding animations may help make this more clear.
Adding other animations around the app may also improve overall user experience.
### Acceptance Criteria (how do I know when I’m done?)
Animation added to make it clear the overlay is an overlay.

### LINKED ISSUE #2646: Update CHANGLOG for v4.1 release
### Vision statement / What / Requirements
v4.1 was released but it's not in the CHANGELOG.

@dnarcese Can you take care of this, as the v4.1 release owner? If so, please assign and move into the sprint. Thanks!

### PR DESCRIPTION
**Pull request by dnarcese**
Also fixes #2646 

This adds an animation to make the focused tile bigger, as well as fading in/out the overlay.

This already got UX sign-off in the issue.
## Checklist

- [x] Confirm the **acceptance criteria** is fully satisfied in the issue(s) this PR will close
- [x] Add thorough **tests** or an explanation of why it does not
animation changes only
- [x] Add a **CHANGELOG entry** if applicable
- [x] Add **QA labels** on the associated issue (not this PR; `qa-ready` or `qa-notneeded`)

### CODE REVIEW

File: app/src/main/res/animator/channel_item_animator.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Baron-Severin) [inline app/src/main/res/animator/channel_item_animator.xml]: Nit: license

File: app/src/main/res/animator/channel_item_animator.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+                android:valueType="floatType"/>
+            <objectAnimator
+                android:duration="@android:integer/config_shortAnimTime"
+                android:propertyName="scaleY"
+                android:valueTo="1.2"
+                android:valueType="floatType"/>
+        </set>
+    </item>
+
+</selector>
```
Comment (Baron-Severin) [inline app/src/main/res/animator/channel_item_animator.xml]: Nit: extra line

File: app/src/main/res/animator/channel_item_animator_not_focused.xml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Baron-Severin) [inline app/src/main/res/animator/channel_item_animator_not_focused.xml]: Nit: license

File: app/src/main/res/animator/channel_item_animator_not_focused.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+                android:valueType="floatType"/>
+            <objectAnimator
+                android:duration="@android:integer/config_shortAnimTime"
+                android:propertyName="scaleY"
+                android:valueTo="1"
+                android:valueType="floatType"/>
+        </set>
+    </item>
+
+</selector>
```
Comment (Baron-Severin) [inline app/src/main/res/animator/channel_item_animator_not_focused.xml]: Nit: extra line

File: app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt [hunk 5]
```diff
@@ -4,6 +4,8 @@
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt]: Interesting note, it looks like this fixes a small visual bug.  I'm not sure what it's number was, but the STR are: have <= 4 tiles remaining in a row -> select any tile index 0 - 2 -> delete tile.  Tile at index 3 clips its border and looks weird.

Presumably one of the `noclip`s added in this PR fixed it.  Nice!

File: app/src/main/res/animator/channel_item_animator.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <item>
+        <set>
+            <objectAnimator
+                android:duration="@android:integer/config_shortAnimTime"
```
Comment (Baron-Severin) [inline app/src/main/res/animator/channel_item_animator.xml]: This is personal opinion, but I feel like this animation might be a little bit too long.  200 is normally a good number, but I wonder if we might want to use 150 or even 100 here.  I do think it's good to err on the side of quicker when it comes to animations.

File: app/src/main/res/values/colors.xml [hunk 7]
```diff
@@ -17,7 +17,7 @@
 
     <color name="onboarding_button_higlight_stroke">@color/photonBlue50</color>
 
-    <color name="browser_overlay_background">@color/photonGrey70</color>
+    <color name="browser_overlay_background">#f21C1B22</color>
```
[the remark this answers is not in the document]
Comment (dnarcese) [reply app/src/main/res/values/colors.xml]: How would the alpha affect performance?

File: app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt [hunk 8]
```diff
@@ -92,13 +94,16 @@ class DefaultChannelAdapter(
                 // We can't use a selector for the tile cardview because we use the focused item to
                 // get the RecyclerView adapter position
                 val focusRingDrawable: Drawable?
+                val animation: StateListAnimator?
                 if (hasFocus) {
                     focusRingDrawable = context.getDrawable(R.drawable.tile_selected_stroke)
+                    animation = AnimatorInflater.loadStateListAnimator(context, R.animator.channel_item_animator)
```
[the remark this answers is not in the document]
Comment (dnarcese) [reply app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt]: This causes it to crash

File: app/src/main/res/drawable/nav_urlbar_background.xml [hunk 9]
```diff
@@ -8,7 +8,7 @@
         <shape
             android:shape="rectangle">
             <corners android:radius="@dimen/urlbar_corner_radius"/>
-            <solid android:color="@color/photonGrey80" />
+            <solid android:color="#32313c" />
```
[the remark this answers is not in the document]
Comment (dnarcese) [reply app/src/main/res/drawable/nav_urlbar_background.xml]: I got the colours from the mocks but I think our colour palette is going to be changing. Maybe I should hold off on any colour changes until they are all changed together.

File: app/src/main/res/values/colors.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/res/values/colors.xml]: In Android, when you draw one view on top of another, you get overdraw* and suffer a performance penalty. When one of those views contains alpha, it's worse: you suffer the same performance penalties as overdraw _and_ you have to compute the new blended value for the pixel (side note: I believe this combining two views where one has alpha is called "compositing"). In general, adding alpha values is a red flag that performance will be affected negatively.

While doing this computation once or a handful of times is fine, afaik, when the WebView is loading pages, it redraws the page over and over again as new elements are loaded. When you combine this with the 60 FPS redraws from animations while moving through items on the homescreen, you're more likely to start dropping frames.

In theory, it's possible this also slows down page load: since one CPU core is more committed to redrawing the page, it can't be used for page load.

I didn't measure frames dropped objectively so I don't know for sure but it felt like we were dropping frames so I wanted to split it out so we can investigate further and see if the trade-off is worthwhile.

\*overdraw is when you draw one color in a pixel and replace it with another color later, invalidating the need for the first pixel to have ever been drawn. If you do this for many pixels (or many times for single pixels), it non-trivially slows down drawing and you can start dropping frames.

File: app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt]: @dnarcese What's the error on the crash?

File: app/src/main/res/drawable/nav_urlbar_background.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/res/drawable/nav_urlbar_background.xml]: @dnarcese Which mocks? Do you mind linking me? I'm not sure I've seen them.

[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/res/drawable/nav_urlbar_background.xml]: > I think our colour palette is going to be changing. Maybe I should hold off on any colour changes until they are all changed together.

Good find! That palette change could be a ways off so it wouldn't hurt to ask Bram if this might be a problem sooner!

[the remark this answers is not in the document]
Comment (dnarcese) [reply app/src/main/res/drawable/nav_urlbar_background.xml]: https://github.com/mozilla-mobile/firefox-tv/issues/2468#issue-461305344

File: app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (dnarcese) [reply app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt]:
```
JNI DETECTED ERROR IN APPLICATION: can't call void android.view.View.setScaleX(float) on null object
in call to CallVoidMethodV
from void android.animation.PropertyValuesHolder.nCallFloatMethod(java.lang.Object, long, float)
```

File: app/src/main/res/values/colors.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (Baron-Severin) [reply app/src/main/res/values/colors.xml]: On the other hand, we have a huge problem with people not seeing the overlay as an overlay (because it... kinda isn't yet).  AFAIK, our major proposals for fixing this are basically adding transparency or splitting the overlay itself.  Either one is going to mean rendering the WV in the background.  We should test, but I wouldn't expect a perceivable perf difference between the two.

TL;DR my guess-that-is-backed-by-zero-data that the perf difference between split/transparent overlay is minimal.  If so, we might need to take one way or another in order to improve our UX.

File: app/src/main/res/layout/home_tile.xml [hunk 10]
```diff
@@ -14,7 +14,8 @@
               android:layout_height="wrap_content"
               android:layout_marginStart="12dp"
               android:focusable="true"
-              android:layout_marginEnd="12dp">
+              android:layout_marginEnd="12dp"
+              android:clipChildren="false">
 
     <LinearLayout
         android:layout_width="@dimen/home_tile_width"
```
[the remark this answers is not in the document]
Comment (dnarcese) [reply app/src/main/res/layout/home_tile.xml]: I was able to remove one, but the other ones are needed because there are several layouts that to be not clipped.

File: app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/java/org/mozilla/tv/firefox/channels/DefaultChannelAdapter.kt]: That's weird. Do you know what the null object is? Is this a bug in the android framework?

File: app/src/main/res/values/colors.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/res/values/colors.xml]: > we have a huge problem with people not seeing the overlay as an overlay

But is this solved by animating the overlay in?

[the remark this answers is not in the document]
Comment (dnarcese) [reply app/src/main/res/values/colors.xml]: I think animating it in helps with this. That way it is easier to tell where it is in relation to the webpage. By fading in, it is clearer that it is a page that is just over the webpage. The transparency will help even more.

File: app/src/main/res/values/configuration.xml [hunk 11]
```diff
@@ -5,4 +5,5 @@
 <resources>
     <!-- We are overriding this value in values-sw600dp -->
 
+    <integer name="channelTileAnimTime">100</integer>
```
Comment (mcomella) [inline app/src/main/res/values/configuration.xml]: It feels weird to put this in `configuration.xml`: it was originally added in Focus to distinguish between configurations, specifically "mobile" and "tablet" and we're not doing that here.

In Android, this would conventionally go in an `integers.xml`. That sounds like a good place to put it to me for now: what do you think?

Also, it seems `values*/configuration.xml` are empty: do you mind removing them while you're here?

Comment (mcomella) [inline app/src/main/res/values/configuration.xml]: I like this timing better: nice! What's interesting is that it's so fast that it's not super noticeable that it's animating (when compared to the translation animations) but it gets serves the intended effect of emphasizing the selected item.

It might be slightly too fast (I think it completes before translation does) but I think this is good enough. :)

**Comment by mcomella** [comment]
I noticed one of the commits is, "Issue #2454: Fix nits". When code spelunking, I think it's helpful to have more descriptive commit messages. For example, if I was bisecting to find a regression and I saw, "Fix nits", it's unclear if that commit could have caused the regression – it could do anything, really! Whereas if the commit is, "Clarify name of channel item animator." or "Encapsulate nav overlay transitions config into helper method", I can more easily determine if it's related. That being said, sometimes it's necessary to break up commits more heavily to accomodate this so if it's way too much work, don't do it but in this case, I think it could be worthwhile.
